containers/src/Data/IntMap/Internal.hs

```diff
@@ -3547,6 +3521,32 @@ binCheckRight _ l Nil = l
 binCheckRight p l r   = Bin p l r
 {-# INLINE binCheckRight #-}
 
+{--------------------------------------------------------------------
+  Branching
+--------------------------------------------------------------------}
+
+data MapMapBranch
+  = ABL  -- ^ A contains B in the left child
+  | ABR  -- ^ A contains B in the right child
+  | BAL  -- ^ B contains A in the left child
+  | BAR  -- ^ B contains A in the right child
+  | EQL  -- ^ A and B have equal prefixes
+  | NOM  -- ^ A and B have prefixes that do not match
+
+mapMapBranch :: Prefix -> Prefix -> MapMapBranch
+mapMapBranch p1 p2 = case compare pw1 pw2 of
+  LT | pw2 <= (pw1 .|. (pw1-1)) -> ABR
+     | (pw2 .&. (pw2-1)) <= pw1 -> BAL
+     | otherwise                -> NOM
+  GT | pw1 <= (pw2 .|. (pw2-1)) -> BAR
+     | (pw1 .&. (pw1-1)) <= pw2 -> ABL
+     | otherwise                -> NOM
+  EQ                            -> EQL
```

Comment:
Moving `ABL` above `BAR` would mean `(pw1 - 1)` is shared between both branches of `pw1 < pw2`, so it can be moved up in the flow.

Reply:
True, didn't think of that.